store/tikv/region_request.go

```diff
@@ -62,6 +62,37 @@ type RegionRequestSender struct {
 	failStoreIDs map[uint64]struct{}
 }
 
+// RegionBatchRequestSender sends BatchCop requests to TiFlash server by stream way.
+type RegionBatchRequestSender struct {
+	RegionRequestSender
+}
+
+// NewRegionBatchRequestSender creates a RegionBatchRequestSender object.
+func NewRegionBatchRequestSender(cache *RegionCache, client Client) *RegionBatchRequestSender {
+	return &RegionBatchRequestSender{RegionRequestSender: RegionRequestSender{regionCache: cache, client: client}}
+}
+
+func (ss *RegionBatchRequestSender) sendReqToAddr(bo *Backoffer, ctxs []copTaskAndRPCContext, req *tikvrpc.Request, timout time.Duration) (resp *tikvrpc.Response, retry bool, err error) {
+	// use the first ctx to send request, because every ctx has same address.
+	ctx := ctxs[0].ctx
+	if e := tikvrpc.SetContext(req, ctx.Meta, ctx.Peer); e != nil {
+		return nil, false, errors.Trace(e)
+	}
+	resp, err = ss.client.SendRequest(bo.ctx, ctx.Addr, req, timout)
+	if err != nil {
+		ss.rpcError = err
+		for _, failedCtx := range ctxs {
+			e := ss.onSendFail(bo, failedCtx.ctx, err)
+			if e != nil {
+				return nil, false, errors.Trace(e)
+			}
+		}
+		return nil, true, nil
+	}
+	// We don't need to process region error or lock error. Because TiFlash will retry by itself.
```

Comment:
What's this comment means? What if some regions in the batch request have been migrated? I prefer to keep retry logic in TiDB.

Reply:
As descriprition in this PR, we decide to put retry logic inside TiFlash. When some regions have been migrated, we can use `client-c` to read remotely. There are two pros:

- It can reduce the cost of rpc communication.
- For some quries, like join that will be implemented in near future, a request means we have to load small table once, this is a heavy operation. But Tiflash can only retry the `datascan + selection` part, which can avoid read small table repeatly,